### PR TITLE
Adds resource BitbucketServerConfig to CloudBuild.

### PR DESCRIPTION
**Pull request by mariomachado94**
Adds resource BitbucketServerConfig to CloudBuild.
- Custom code is added to support connected_connected repositories field via separate API calls

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_cloudbuild_bitbucket_server_config`
```

### CODE REVIEW

File: mmv1/products/cloudbuild/api.yaml [hunk 1]
```diff
@@ -25,6 +25,138 @@ apis_required:
...
+              required: true
+              description: |
+                Identifier for the project storing the repository.
+            - !ruby/object:Api::Type::String
+              name: 'repoSlug'
+              required: true
+              description: |
+                Identifier for the repository.
+      - !ruby/object:Api::Type::String
+        name: 'peeredNetwork'
```
Comment (trodge) [inline mmv1/products/cloudbuild/api.yaml:150]: Would it be possible to add a test that uses the `peered_network` and `ssl_ca` fields?

**Review by trodge — CHANGES_REQUESTED** [review]

File: mmv1/products/cloudbuild/api.yaml (hunk 1 above)
Comment (mariomachado94) [reply mmv1/products/cloudbuild/api.yaml:150]: I attempted this, and ended up only adding an extra example while skipping test generation. 

tl;dr -  I would need to use `google_compute_network`, and it seems to be unable to delete itself.

You can see in the peered_network example I added that I need to use the compute api and service networking apis to create dependant resources. When testing manually, things work until `terraform destroy`. It seems that while using the `google_compute_network`, firewall rules are created for that network, which then causes issues on deletion. The test fails to clean up the resources.

Can we move on with just the extra docs?

Comment (trodge) [reply mmv1/products/cloudbuild/api.yaml:150]: Could you try using `BootstrapSharedTestNetwork`? There is an example [here](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/vertexai/terraform.yaml#L58).

**Review by trodge — COMMENTED** [review]
LGTM, adding Megan for a second review.

File: mmv1/products/cloudbuild/api.yaml [hunk 2]
```diff
@@ -28,6 +28,138 @@ apis_required:
...
+          Immutable. API Key that will be attached to webhook. Once this field has been set, it cannot be changed.
+          If you need to change it, please create another BitbucketServerConfig.
+      - !ruby/object:Api::Type::Array
+        name: 'connectedRepositories'
+        description: |
+          Connected Bitbucket Server repositories for this config.
+        item_type: !ruby/object:Api::Type::NestedObject
+          name: 'BitbucketServerRepositoryId'
+          description: |
+            BitbucketServerRepositoryId identifies a specific repository hosted on a Bitbucket Server.
```
Comment (shuyama1) [inline mmv1/products/cloudbuild/api.yaml:140]: I think we can remove these two lines. Don't think they will get generated.

File: mmv1/products/cloudbuild/api.yaml [hunk 3]
```diff
@@ -28,6 +28,138 @@ apis_required:
...
+          Output only. UUID included in webhook requests. The UUID is used to look up the corresponding config.
+      - !ruby/object:Api::Type::String
+        name: 'apiKey'
+        required: true
+        input: true
+        description: |
+          Immutable. API Key that will be attached to webhook. Once this field has been set, it cannot be changed.
+          If you need to change it, please create another BitbucketServerConfig.
+      - !ruby/object:Api::Type::Array
+        name: 'connectedRepositories'
```
Comment (shuyama1) [inline mmv1/products/cloudbuild/api.yaml:134]: I understand that we probably not able to add tests for this field in our CI. Got a quick question regarding your local test:
1. Have you tested the update functionality of this field?

File: mmv1/products/cloudbuild/api.yaml [hunk 4]
```diff
@@ -28,6 +28,138 @@ apis_required:
...
+            description: |
+              Immutable. The resource name for the webhook secret's secret version. Once this field has been set, it cannot be changed.
+              If you need to change it, please create another BitbucketServerConfig.
+      - !ruby/object:Api::Type::Time
+        name: 'createTime'
+        output: true
+        description: |
+          Time when the config was created.
+      - !ruby/object:Api::Type::String
+        name: 'username'
```
Comment (shuyama1) [inline mmv1/products/cloudbuild/api.yaml:117]: Just want to confirm the behavior of this field. If we don't specify this field explicitly, will the API return a value for this field?

File: mmv1/templates/terraform/examples/cloudbuild_bitbucket_server_config_peered_network.tf.erb [hunk 5]
```diff
@@ -0,0 +1,50 @@
+terraform {
+  required_providers {
+    google = {}
+  }
+}
+
+provider "google" {}
```
Comment (shuyama1) [inline mmv1/templates/terraform/examples/cloudbuild_bitbucket_server_config_peered_network.tf.erb:7]: any reason we need to set these explicitly?

File: mmv1/products/cloudbuild/api.yaml (hunk 3 above)
Comment (mariomachado94) [reply mmv1/products/cloudbuild/api.yaml:134]: I did. I had to be sure to add custom code in the preupdate to remove this field from the field mask, as well as to then update it with custom code. Then made sure to test edge cases, 0 -> 1, many as well as update many, 1 -> 0, as well as a few other permutations.

File: mmv1/products/cloudbuild/api.yaml (hunk 4 above)
Comment (mariomachado94) [reply mmv1/products/cloudbuild/api.yaml:117]: No, it needs this field from the user, it will return an error if not specified (it is the bitbucket server user name).

File: mmv1/products/cloudbuild/api.yaml (hunk 2 above)
Comment (mariomachado94) [reply mmv1/products/cloudbuild/api.yaml:140]: Ah, yes, you are right. I assume you mean name as well? It looks like name and description are not generated. Removing both.

File: mmv1/templates/terraform/examples/cloudbuild_bitbucket_server_config_peered_network.tf.erb (hunk 5 above)
Comment (mariomachado94) [reply mmv1/templates/terraform/examples/cloudbuild_bitbucket_server_config_peered_network.tf.erb:7]: Nope. Good catch, removing them.

File: mmv1/products/cloudbuild/api.yaml [hunk 6]
```diff
@@ -28,6 +28,138 @@ apis_required:
...
+            description: |
+              The resource name for the read access token's secret version.
+          - !ruby/object:Api::Type::String
+            name: 'webhookSecretVersionName'
+            required: true
+            input: true
+            description: |
+              Immutable. The resource name for the webhook secret's secret version. Once this field has been set, it cannot be changed.
+              If you need to change it, please create another BitbucketServerConfig.
+      - !ruby/object:Api::Type::Time
```
[the remark this answers is not in the document]
Comment (mariomachado94) [reply mmv1/products/cloudbuild/api.yaml:111]: Okay, removing createTime.

File: mmv1/products/cloudbuild/api.yaml (hunk 4 above)
Comment (shuyama1) [reply mmv1/products/cloudbuild/api.yaml:112]: Sorry, I missed `required: true` setting somehow and thought it was an optional field, and that's why I asked the question. Thanks for confirming though!

**Review by shuyama1 — COMMENTED** [review]
Only some nitpicks and questions. Sorry for missing some of these in the previous review.

File: mmv1/templates/terraform/pre_update/cloudbuild_bitbucketserver_config.go.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    if field == "connectedRepositories" {
+        updateMask = append(updateMask[:i], updateMask[i+1:]...)
+        break
+    }
+}
+// reconstruct url
+url, err = replaceVars(d, config, "{{CloudBuildBasePath}}projects/{{project}}/locations/{{location}}/bitbucketServerConfigs/{{config_id}}")
+if err != nil {
+    return err
+}
```
[the remark this answers is not in the document]
Comment (mariomachado94) [reply mmv1/templates/terraform/pre_update/cloudbuild_bitbucketserver_config.go.erb:12]: I'm pretty sure I thought the same when I first added this code. But at this point the value of url already has the bad update mask included by a previous 'addQueryParams' call in the generated code. I have to recreate the url variable before calling addQueryParams.

File: mmv1/products/cloudbuild/terraform.yaml [hunk 8]
```diff
@@ -14,6 +14,38 @@
...
+      - !ruby/object:Provider::Terraform::Examples
+        name: "cloudbuild_bitbucket_server_config_peered_network"
+        primary_resource_id: "bbs-config-with-peered-network"
+        vars:
+          network_name: "vpc-network"
+        test_vars_overrides:
+          network_name: 'BootstrapSharedTestNetwork(t, "peered-network")'
+    properties:
+      bitbucketServerConfigId: !ruby/object:Overrides::Terraform::PropertyOverride
+        name: 'config_id'
```
[the remark this answers is not in the document]
Comment (mariomachado94) [reply mmv1/products/cloudbuild/terraform.yaml:41]: If i remember correctly, I was thinking the original name was a bit verbose from the user's perspective: google_cloudbuild_bitbucket_server_config.myconfig.bitbucket_server_config_id vs google_cloudbuild_bitbucket_server_config.myconfig.config_id

and I think just `id` is already automatically generated if i remember correctly.

File: mmv1/templates/terraform/pre_update/cloudbuild_bitbucketserver_config.go.erb (hunk 7 above)
[the remark this answers is not in the document]
Comment (shuyama1) [reply mmv1/templates/terraform/pre_update/cloudbuild_bitbucketserver_config.go.erb:12]: Thanks for confirming! Let's keep it then, since you've already tested it!

File: mmv1/products/cloudbuild/terraform.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (shuyama1) [reply mmv1/products/cloudbuild/terraform.yaml:41]: Got it! So the rename is just for simplicity. And yes, you're right, `id` is the generated identifier used across all resources.

**Review by shuyama1 — APPROVED** [review]
Thank you for the work!

**Comment by shuyama1** [comment]
Back to @trodge for final approval!

**Review by trodge — APPROVED** [review]
